google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigquery.storage.v1alpha2;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.Schema;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.testing.RemoteBigQueryHelper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Descriptors;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class SchemaCompact {
+  private BigQuery bigquery;
+  private static SchemaCompact compact;
+  private static String tablePatternString = "projects/([^/]+)/datasets/([^/]+)/tables/([^/]+)";
+  private static Pattern tablePattern = Pattern.compile(tablePatternString);
+
+  private SchemaCompact(BigQuery bigquery) {
+    this.bigquery = bigquery;
+  }
+
+  public static SchemaCompact getInstance() {
+    if (compact == null) {
+      RemoteBigQueryHelper bigqueryHelper = RemoteBigQueryHelper.create();
+      compact = new SchemaCompact(bigqueryHelper.getOptions().getService());
+    }
+    return compact;
+  }
+
+  @VisibleForTesting
+  public static SchemaCompact getInstance(BigQuery bigquery) {
+    return new SchemaCompact(bigquery);
+  }
+
+  TableId getTableId(String tableName) {
+    Matcher matcher = tablePattern.matcher(tableName);
+    if (!matcher.matches() || matcher.groupCount() != 3) {
+      throw new IllegalArgumentException("Invalid table name: " + tableName);
+    }
+    return TableId.of(matcher.group(1), matcher.group(2), matcher.group(3));
+  }
+
+  /**
+   * Checks if the userSchema is compatible with the table's current schema for writing.
+   *
+   * @param client
+   * @param tableName
+   * @param userSchema
+   */
+  public void check(String tableName, Descriptors.Descriptor userSchema) {
+    com.google.cloud.bigquery.Table table = bigquery.getTable(getTableId(tableName));
+    Schema schema = table.getDefinition().getSchema();
+    // TODO: We only have very limited check here. More checks to be added.
+    if (schema.getFields().size() != userSchema.getFields().size()) {
```

Comment:
This seems to be a very barebone check to validate if userSchema is compatible w current schema. Could we at least check field type and field mode? The reason is I see that this check is used in WriterCache.

Reply:
Yes, I don't have time to fully implement the check, more checks will be added. As of now, it just prevents very random input, which, without check, would cause a stream to be created (which will take a day to expire). The more check we have, the happier backend would be. I am planning to have my intern to finish this function.